### PR TITLE
PR#66 (renderTourComments | add @dataProvider method)

### PR DESCRIPTION
**Pull request by rgadon107**
## PR Summary

This PR includes one unit and integration test each for the function `render_tour_comments` in the `tours` plugin.  The function gets and echos the post meta `tour_comments` into the HTML view displayed in the title of each past tour.   The unit and integration tests each employ a @dataProvider method to pass test data to their respective test methods.

The relative file path to the function within the plugin is:

> `/tours/src/meta.php`.

### CODE REVIEW

File: plugins/tours/tests/phpunit/integration/renderTourComments.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	/**
+	 * @dataProvider addTestData
+	 */
+	public function test_should_echo_meta_key_values_when_postmeta_exists( $post_data, $meta ) {
+		// Create and get the $tour_id using WordPress' factory method.
+		$tour_id = $this->factory->post->create( $post_data );
+
+		// Add post_meta to the database so we can call it.
+		add_post_meta( $tour_id, 'tour_comments', $meta['tour_comments']
+		);
```
Comment (hellofromtonya) [inline plugins/tours/tests/phpunit/integration/renderTourComments.php:33]: Instead of hard coding the meta, allow for flexibility by registering the meta that's passed from the test data:

```php
foreach( $meta as $key => $value ) {
	add_post_meta( $tour_id, $key, $value );
}
```

This approach gives you flexibility to add more test data edge cases and scenarios without having to modify the test method's code.

File: plugins/tours/tests/phpunit/integration/renderTourComments.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+		$expected = (string) get_post_meta( $tour_id, 'tour_comments', true );
```
Comment (hellofromtonya) [inline plugins/tours/tests/phpunit/integration/renderTourComments.php:35]: This is not what's expected. Right? The expected value should be passed from the test data. Why? The actual expected value depends upon what is being tested from the test data's scenario. Pass it from the test data.

Huh?

Let's look at the code:

https://github.com/rgadon107/cornerstone/blob/b12d9c8aa0593a6a4a6d7af418e2a78c8234a1b3/plugins/tours/src/meta.php#L38-L46

What do you see? What are the scenarios?

1. If the post meta is empty, then the expected behavior is: the function bails out and returns void/null.
2. If there's post meta, then it returns it.
3. Notice the post meta is escaped through `esc_html()`. You want to test that it gets escaped too.

So there are 3 different scenarios to test. And you'll want to pass in different types of "tour comments" to validate it works as you expect.

Does that make sense?

File: plugins/tours/tests/phpunit/integration/renderTourComments.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+		// Run the output buffer to fire the callback and return the output.
+		ob_start();
+		render_tour_comments( $tour_id );
+		$actual = ob_get_clean();
+
+		$this->assertSame( $expected, $actual );
+
+		// Clean up database.
+		delete_post_meta( $tour_id, 'tour_comments' );
```
Comment (hellofromtonya) [inline plugins/tours/tests/phpunit/integration/renderTourComments.php:45]: Move the `delete_post_meta()` and any clean up code to the `tearDown()` method.

Why? If the assertion fails, the test method bails out and does not run the code below the failed assertion. That means this line of code does not run if that happens. But by moving it to `tearDown()`, it will always run.

How can you do that? The `tour_id` and meta keys to delete are passed from the test data. How can the `tearDown()` know which ones to delete? Think about it. What can you do? Assign each to a property to allow methods to access this test scenarios' values. 

```php
public function test_should_echo_meta_key_values_when_postmeta_exists( $post_data, $meta ) {
	// Create and get the $tour_id using WordPress' factory method.
	$this->tour_id = $this->factory->post->create( $post_data );
	$this->meta = $meta;
```

Give it a try.

File: plugins/tours/tests/phpunit/unit/renderTourComments.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	/**
+	 * @dataProvider addTestData
+	 */
+	public function test_should_echo_meta_key_values_when_postmeta_exists( $tour_id, $postmeta ) {
+		Functions\expect( 'get_post_meta' )
+			->once()
+			->with( $tour_id, 'tour_comments', true )
+			->andReturn( $postmeta );
+
+		$this->expectOutputString( $tour_id );
```
Comment (hellofromtonya) [inline plugins/tours/tests/phpunit/unit/renderTourComments.php:43]: This is not what's expected. The tour ID isn't returned, but rather escaped tour comments' value or HTML. Look at the source code:

https://github.com/rgadon107/cornerstone/blob/b12d9c8aa0593a6a4a6d7af418e2a78c8234a1b3/plugins/tours/src/meta.php#L38-L46

Let's build a unit test that matches the test data that's given to it. How?

```php
public function test_should_echo_meta_key_values_when_postmeta_exists( $tour_id, $meta, $expected ) {
	Functions\expect( 'get_post_meta' )
		->once()
		->with( $tour_id, 'tour_comments', true )
		->andReturn( $meta );

	if ( ! is_null( $expected ) {
		$this->expectOutputString( $expected );
	}

	render_tour_comments( $tour_id );
}

You'll need to add the expected value or HTML in the test data and update the scenarios.
